### PR TITLE
Report stats using fewer messages (#68)

### PR DESCRIPTION
**Pull request by galgeek**
Hi, @whimboo!

Here’s a quick but incomplete fix for issue #68, accumulating active users stats and active hours stats each on a single line. This produces a readable stats report using 8 messages plus whatever additional messages node-irc splits off for >512 character lines of active users and active hours.

node-irc splits message not only at 512 characters but also at \n. It's challenging to produce a readable report using the minimum number of messages without using \n.

Maybe this is good enough?

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -438,28 +438,31 @@ var Stats = function() {};
 Stats.prototype.generateStats = function(metrcs, from) {
   var keys = Object.keys(metrcs);
   var what = Object.prototype.toString;
+  var report ="";
+
   for (var i = 0; i < keys.length; i++) {
     if (what.call(metrcs[keys[i]]).search('Array') > 0) {
-      client.say(from, keys[i] + ":  " + metrcs[keys[i]].join(", "));
+      report = report + "\n" + keys[i] + ":  " + metrcs[keys[i]].join(", ");
```
Comment (whimboo) [inline bot.js]: I would suggest to add the backslash to the end of the lines but not to the beginning.

File: bot.js [hunk 2]
```diff
@@ -438,28 +438,31 @@ var Stats = function() {};
...
     } else {
       if (keys[i] == "activeUsers") {
         var speakers = Object.keys(metrcs.activeUsers);
         var speakersTotal = speakers.length;
-        client.say(from, "The following " + speakersTotal + " people were active in the channel: ");
+        report = report + "\nThe following " + speakersTotal + " people were active in the channel:  ";
         for (var t = 0; t < speakersTotal; t++) {
-          client.say(from, speakers[t] + ": " + metrcs.activeUsers[speakers[t]]);
+          report = report + speakers[t] + ": " + metrcs.activeUsers[speakers[t]] + "  *  ";
```
Comment (whimboo) [inline bot.js]: Hm, where do we add those new lines here between the individual hours?

**Comment by whimboo** [comment]
If you can bring this code up with the testdaybot I would like to have a test with.

File: bot.js (hunk 2 above)
Comment (galgeek) [inline bot.js]: We don't add new lines between individual hours, to shorten the stats report.

File: bot.js (hunk 1 above)
Comment (galgeek) [inline bot.js]: It's a little more readable to put the \n at line end, but it requires a couple more lines of code, since we're no longer adding \n between active users and active hours. I've updated the PR.

**Comment by galgeek** [comment]
Hi, @ashughes1 and @whimboo!

I've updated this PR per Henrik’s January 26 comments, and it's running now in #testdaybotTest

File: bot.js (hunk 2 above)
Comment (whimboo) [inline bot.js]: It doesn't shorten the report. Instead it adds a couple more characters due to the blanks and start separator you are using here. A `\n` only needs a single char compared to 3 as I see above. So please keep separating it with it. 

Maybe it would be good to print a JSON string for all, which could easily be imported into another system for analysis later.

Comment (galgeek) [inline bot.js]: It's true that this approach includes more characters, and at the same time, it uses fewer lines and fewer messages to send the same information, while the spacing makes the reporting still readable. I have changed the spacing in the updated PR to just ", " which is probably readable enough.

It would be simple to add a command to print the JSON, and good to remember that JSON > 512 characters will get split across messages. I'd guess if the :stats output from Friday's test day was sent using many too many messages that the JSON version would be split over multiple messages, well, at least two.

**Comment by galgeek** [comment]
Hello, @whimboo and @ashughes1!

I've further refined the stats output, and I'll plan to leave the code running  for a while in  #testdaybotTest once I return home later today. In particular, stats output will be briefer when a test day is active, since it'll no long bother to print the topic, etherpad, or start and end times then, when they'll be widely known and easy to find otherwise. See my line notes above for minor formatting updates.

If you'd prefer only to output the JSON, keeping in mind that it may be split over two or more messages, let me know.

File: bot.js [hunk 3]
```diff
@@ -438,28 +438,36 @@ var Stats = function() {};
...
+      report = report + keys[i] + ":  " + metrcs[keys[i]].join(", ") + "\n";
     } else {
       if (keys[i] == "activeUsers") {
         var speakers = Object.keys(metrcs.activeUsers);
         var speakersTotal = speakers.length;
-        client.say(from, "The following " + speakersTotal + " people were active in the channel: ");
-        for (var t = 0; t < speakersTotal; t++) {
-          client.say(from, speakers[t] + ": " + metrcs.activeUsers[speakers[t]]);
+        report = report + "The following " + speakersTotal + " people were active in the channel:  ";
+        for (var t = 0; t < speakersTotal-1; t++) {
```
Comment (whimboo) [inline bot.js]: Why the newly introduced `-1` here? You will miss the very last speaker in the list.

File: bot.js [hunk 4]
```diff
@@ -438,28 +438,36 @@ var Stats = function() {};
...
+          report = report + speakers[t] + ": " + metrcs.activeUsers[speakers[t]] + ", ";
         }
+        report = report + speakers[t] + ": " + metrcs.activeUsers[speakers[t]] + "\n";
```
Comment (whimboo) [inline bot.js]: Oh, I see it's because of the final separator. I wouldn't mind if a colon stays at the end of the line. Or make an `var sep = (t == speakersTotal - 1) ? "\n" : ", "`.

File: bot.js [hunk 5]
```diff
@@ -438,28 +438,36 @@ var Stats = function() {};
...
       } else if (keys[i] == "hourUTC") {
-        client.say(from, "The following hours (UTC) were active in the channel: ");
+        report = report + "The following hours (UTC) were active in the channel:  ";
         var speakers = Object.keys(metrcs.hourUTC);
-        for (var t = 0; t < speakers.length; t++) {
-          client.say(from, speakers[t] + ": " + metrcs.hourUTC[speakers[t]]);
+        for (var t = 0; t < speakers.length-1; t++) {
+          report = report + speakers[t] + ": " + metrcs.hourUTC[speakers[t]] + ", ";
         }
+        report = report + speakers[t] + ": " + metrcs.hourUTC[speakers[t]] + "\n";
```
Comment (whimboo) [inline bot.js]: Same as above.

File: bot.js [hunk 6]
```diff
@@ -438,28 +438,36 @@ var Stats = function() {};
...
       } else {
-        client.say(from, keys[i] + ": " + metrcs[keys[i]]);
+        // report most metrics' string properties only when testday is not active
+        if ((!testDay.active) || (testDay.active && keys[i] === "optOutTotal")) {
```
Comment (whimboo) [inline bot.js]: This is changing the logic of the metrics output. This should not be part of this PR and clearly needs a discussion if this is really wanted. I wouldn't care for it right now, so just leave it as it was before.

**Comment by galgeek** [comment]
Hi, @whimboo and @ashughes1!

Henrik, thanks for your feedback! I've updated the PR.

I've left this code merged with the firebot updates running in #testdaybotTest. I've added you both, only in the code that's running, as well as Florin, along with firebot, as bug senders, so you can check that update, too, with a message like "Bug 15894 - Test mozilla/testdaybot updates". 

Check stats with a PM to _TestDayBotTestBot of ":stats".

**Comment by whimboo** [comment]
I tested the command against the bot but I got back 10 messages. I do not think that it is running with the latest code changes?

**Comment by galgeek** [comment]
Hi, @whimboo!

I get only 8 messages from the :stats command.

I got 2 more at the same time, the first time I ran :stats just now, since it was the first message I sent of the "current testday", welcoming me and letting me know I could opt out of tracking, but those are not part of the :stats output.

**Comment by whimboo** [comment]
With the current instance of _TestDayBotTestBot running I get 10 messages. Could it be that every `\n` gets converted into a new message?

**Comment by whimboo** [comment]
Hm, I have tested with another IRC client as sender but it is the same. But when I send from my local client to the other one it works. So I assume this is a setting in HexChat.

**Comment by whimboo** [comment]
So all is fine here. I squashed the commits, and merged the PR as https://github.com/mozilla/testdaybot/commit/664303bc9d74423ba425e360423bb2b1f8babe7a.
